### PR TITLE
[Filebeat][CometD] Add CometD input for Salesforce connector

### PR DESCRIPTION
**Pull request by kush-elastic**
- Enhancement

## What does this PR do?
Real-time logs from the Salesforce Streaming API can be collected using the cometD input. For example, live Login & Logout related logs.

Right now with this PR, when the cometd input is enabled, you can start seeing the authentication log messages that are retrieved after successful authentication using the credentials provided as a part of the configuration. You will also be able to see the real-time data being collected from the Salesforce Steaming API.
For configuration:
```
- type: cometd
  channel_name: /event/LoginEventStream
  auth.oauth2:
    client.id: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    client.secret: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
    token_url: https://login.salesforce.com/services/oauth2/token
    user: abc.xyz@mail.com
    password: P@$$W0₹D
```

## Why is it important?
- It is a prerequisite for the Salesforce Observability Connector for the collection of real-time data from the Streaming APIs.
- With all the different logs in cometD from different services, it will be good to have a dedicated Filebeat input to retrieve the same using cometD.

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [X] I have made corresponding change to the default configuration files
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

### CODE REVIEW

**Comment by kush-elastic** [comment]
Hey @cmacknz, this is the new PR for CometD input.

**Comment by cmacknz** [comment]
If I check out this PR locally and run the tests it is still flaky for me. For flaky tests you need to run them more then once, I usually put them in a shell while loop with a go test flag that will disable test caching like `-count 1`:

```go
~/go/src/github.com/elastic/beats pr/kush-elastic/31492-1:salesforce_cometd_input ?1 ···· 7s 11:49:18 AM
❯ while go test -count 1 ./x-pack/filebeat/input/cometd/...; do :; done
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        5.426s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        5.436s
--- FAIL: TestMultiInput (5.00s)
    input_test.go:339: 
                Error Trace:    input_test.go:339
                Error:          Received unexpected error:
                                not able to get events.
                Test:           TestMultiInput
FAIL
FAIL    github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        5.434s
FAIL
```

**Comment by cmacknz** [comment]
Also odd, in a Jenkins run where the build is green I can see this test actually failed. Looking at the console text for https://beats-ci.elastic.co/job/Beats/job/beats/job/PR-31492/5/ I can see the following:

```go
[2022-05-03T10:24:36.970Z] === Failed
[2022-05-03T10:24:36.970Z] === FAIL: x-pack/filebeat/input/cometd TestMultiInput (5.00s)
[2022-05-03T10:24:36.970Z]     input_test.go:339: 
[2022-05-03T10:24:36.970Z]         	Error Trace:	input_test.go:339
[2022-05-03T10:24:36.970Z]         	Error:      	Received unexpected error:
[2022-05-03T10:24:36.970Z]         	            	not able to get events.
[2022-05-03T10:24:36.970Z]         	Test:       	TestMultiInput
[2022-05-03T10:24:36.970Z] 
[2022-05-03T10:24:36.970Z] DONE 832 tests, 7 skipped, 1 failure in 190.543s
[2022-05-03T10:24:36.970Z] Error: go test returned a non-zero value: exit status 1
[2022-05-03T10:24:36.978Z] [Pipeline] }
[2022-05-03T10:24:36.979Z] ERROR: script returned exit code 1
[2022-05-03T10:24:36.979Z] Retrying
```

I think Jenkins might have some built in retries here

**Comment by kush-elastic** [comment]
Thanks for the details, I looked into those tests. there were some problems with immediate starting of all the tests. and cause of server issues. I have updated those. let's see if that works for Jenkins or not.

Here's what you suggested.
image

**Comment by cmacknz** [comment]
Thanks, I left it in the background for a while and managed to get it to fail:

```
❯ while go test -count 1 ./x-pack/filebeat/input/cometd/...; do :; done
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.458s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.427s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.443s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.457s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.452s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.439s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.397s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.446s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.446s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.449s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.395s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.403s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.412s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.440s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.455s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.377s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.445s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.471s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.375s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.412s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.449s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.453s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.443s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.397s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.440s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.458s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.437s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.454s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.442s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.448s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.427s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.462s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.442s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.440s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.442s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.435s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.453s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.458s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.435s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.477s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.447s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.434s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.420s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.484s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.408s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.403s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.441s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.411s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.448s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.440s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.452s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.443s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.448s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.435s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.459s
ok      github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.428s
--- FAIL: TestMultiInput (2.01s)
    input_test.go:310: 
                Error Trace:    input_test.go:310
                Error:          Received unexpected error:
                                not able to get events.
                Test:           TestMultiInput
FAIL
FAIL    github.com/elastic/beats/v7/x-pack/filebeat/input/cometd        2.435s
FAIL
```

**Comment by kush-elastic** [comment]
> Thanks, I left it in the background for a while and managed to get it to fail:

Thanks, @cmacknz! Do you have anything on top of your head which can help us resolve this? I am also trying some things out.

**Comment by cmacknz** [comment]
No, but I'll look closer at the code. I have had tests like this that require 10+ minutes of continuous running to see failures, so it could be worse :)

**Comment by kush-elastic** [comment]
Hey @cmacknz,
I looked into this and found out that problem was in http test server.
here's solution:
```
if called < uint64(inputType) {
	atomic.AddUint64(&called, 1)
	_, _ = w.Write([]byte(`[{"data": {"payload": {"CountryIso": "IN"}, "event": {"replayId":1234}}, "channel": "channel_name"}]`))
} else {
	_, _ = w.Write([]byte(`{}`))
}
```

I was expecting empty response from server after few events. after receiving those I should not send more events in response based on condition. so what was missing is `{}`.

**Comment by kush-elastic** [comment]
Hey @cmacknz, The E2E tests are again failing.
Is there any specific reason for this to failures?

Still I am re-triggering CI with `/package`.

File: x-pack/filebeat/input/cometd/input_test.go [hunk 1]
```diff
@@ -170,20 +168,22 @@ func TestInputStop_Wait(t *testing.T) {
...
 	wg.Add(1)
+	waitForConnections.Add(1)
 	go func() {
 		require.Equal(t, 1, bay.GetConnectedCount()) // current open channels count should be 1
 		event := <-eventsCh
 		assertEventMatches(t, expected, event) // wait for single event
 		wg.Done()
 		time.Sleep(100 * time.Millisecond)           // let input.Stop() be executed.
 		require.Equal(t, 0, bay.GetConnectedCount()) // current open channels count should be 0
+		waitForConnections.Done()
```
Comment (cmacknz) [inline x-pack/filebeat/input/cometd/input_test.go:181]: Do you need a separate waitgroup for this? Can you just move the `wg.Done()` call above here and get the same result?

File: x-pack/filebeat/input/cometd/input_test.go [hunk 2]
```diff
@@ -363,9 +329,11 @@ func oauth2Handler(w http.ResponseWriter, r *http.Request) {
 		_, _ = w.Write([]byte(`[{"ext":{"replay":true,"payload.format":true},"minimumVersion":"1.0","clientId":"client_id","supportedConnectionTypes":["long-polling"],"channel":"/meta/handshake","version":"1.0","successful":true}]`))
 		return
 	case `{"channel": "/meta/connect", "connectionType": "long-polling", "clientId": "client_id"} `:
-		if called < 2 {
+		if called < uint64(inputType) {
```
Comment (cmacknz) [inline x-pack/filebeat/input/cometd/input_test.go:332]: Can you rename `inputType` to something that better describes what it is? Maybe `expectedHTTPEventCount`?

File: x-pack/filebeat/input/cometd/input_test.go (hunk 1 above)
Comment (kush-elastic) [reply x-pack/filebeat/input/cometd/input_test.go:181]: Yes, I need that to wait for entire test to finishes. `wg.Done()` is to wait for event to get and `waitForConnections.Done()` is to see if all the connections are closed after `input.wait()` is called.

File: x-pack/filebeat/input/cometd/input_test.go (hunk 2 above)
Comment (kush-elastic) [reply x-pack/filebeat/input/cometd/input_test.go:332]: Sure.

File: x-pack/filebeat/input/cometd/input_test.go (hunk 1 above)
Comment (cmacknz) [reply x-pack/filebeat/input/cometd/input_test.go:181]: Can you give the original `wg` variable a more descriptive name then, to make it clear why it needs to stay?

Comment (kush-elastic) [reply x-pack/filebeat/input/cometd/input_test.go:181]: Sure, maybe I can give it a name like `waitEventCollection`.

Comment (kush-elastic) [reply x-pack/filebeat/input/cometd/input_test.go:181]: or I guess `waitForEventCollection` would be better.

**Review by cmacknz — APPROVED** [review]

**Comment by kush-elastic** [comment]
Hey @cmacknz,
I everything other then E2E test are passing. Can we merge this PR? If so, then can you merge it as we don't rights to do it?
Or is there something else we need to do?

**Comment by mtojek** [comment]
I synced with @kush-elastic regarding the status of E2E tests. We agreed to merge based on the same justification as [here](https://github.com/elastic/beats/pull/30089/#issuecomment-1114953268).

**Pull request merged**